git/ignore.go

```diff
@@ -0,0 +1,275 @@
+package git
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+// An ignore pattern declared in a pattern file (e.g. .gitignore) or provided as an input.
+type IgnorePattern struct {
+	Pattern string // The pattern to match, relative to the scope
+	Source  File   // The path to the pattern file where this pattern is found, relative path for work dir resources, otherwise absolute path or blank for one that was provided as input to dgit
+	Scope   File   // The work directory relative scope that this pattern applies
+	LineNum int    // The line number within the source pattern file where this pattern is located
+}
+
+// Represents a match (or non-match) of a particular path name against a set of ignore patterns
+type IgnoreMatch struct {
+	IgnorePattern      // A match has all of the pattern information, or empty pattern if no match was found
+	PathName      File // The provided path name that was checked for the ignore
+}
+
+// Returns the standard representation of an ignore match (or non-match)
+func (im IgnoreMatch) String() string {
+	return fmt.Sprintf("%s:%s:%s\t%s", im.Source, im.LineString(), im.Pattern, im.PathName)
+}
+
+func (im IgnoreMatch) LineString() string {
+	lineNum := ""
+	if im.LineNum != 0 {
+		lineNum = strconv.Itoa(im.LineNum)
+	}
+	return lineNum
+}
+
+func (ip IgnorePattern) Matches(ignorePath string, isDir bool) bool {
+	if !strings.HasPrefix(ignorePath, ip.Scope.String()) {
+		return false
+	}
+
+	rel, err := filepath.Rel("/"+ip.Scope.String(), "/"+ignorePath)
+
+	// Not in scope of this pattern
+	if err != nil || strings.HasPrefix(rel, ".") {
+		return false
+	}
+
+	return matchesGlob("/"+rel, isDir, ip.Pattern)
+}
+
+// Returns the standard ignores (.gitignore files and various global sources) that
+//  should be used to determine whether the provided files are ignored. These patterns
+//  can be used with these files with ApplyIgnore() to find whether each file is ignored
+//  or not.
+func StandardIgnorePatterns(c *Client, paths []File) ([]IgnorePattern, error) {
+	alreadyParsed := make(map[string]bool)
+	finalPatterns := []IgnorePattern{}
+
+	for _, path := range paths {
+		abs, err := filepath.Abs(path.String())
+		if err != nil {
+			return nil, err
+		}
+
+		// Let's check that this path is in the git work dir first
```

Comment:
This check seems useful enough that it might make sense to make it a method on File rather than inline in the loop.